File: ruby/Exercise2/CodeToRefactor.rb
require 'date'

module Review
    class People
    @under_16 = DateTime.now - (15 * 365)

    class << self
      def under_16
        @under_16
      end

      private :under_16
    end

    attr_reader :name, :dob

    def initialize(name, dob = nil)
      if dob.nil?
        dob = self.class.send(:under_16)
      end
      @name = name
      @dob = dob
    end
  end

  class BirthingUnit
    # MaxItemsToRetrieve
    def initialize
      @people = []
    end

    # GetPeoples
    # @param j
    # @return Array<Object>
    def get_people(i)
      i.times do |j|
        begin
          # Creates a dandon Name
          name = ""
          if rand(0..1) == 0
            name = "Bob"
          else
            name = "Betty"
          end
          # Adds new people to the list
          @people << People.new(name, DateTime.now - (rand(18..85) * 356))
        rescue => e
          # Dont think this should ever happen
          raise "Something failed in user creation"
        end
      end
      @people
    end

    private

    def get_bobs(older_than_30)
      older_than_30 ? @people.select { |x| x.name == "Bob" && x.dob >= (DateTime.now - (30 * 356)) } : @people.select { |x| x.name == "Bob" }
    end

    public

    def get_married(p, last_name)
      return p.name if last_name.include?("test")
      
      if (p.name.length + last_name).length > 255
        (p.name + " " + last_name)[0, 255]
      end

      p.name + " " + last_name 
    end
  end
end

File: typescript/Exercise2/CodeToRefactor.ts
namespace Review 
{
    export class People {
        private static Under16: Date = new Date(Date.now() - (15 * 365 * 24 * 60 * 60 * 1000));
        public Name: string;
        public DOB: Date;

        constructor(name: string);
        constructor(name: string, dob: Date);
        constructor(name: string, dob?: Date) {
            if (dob === undefined) {
                this.Name = name;
                this.DOB = People.Under16;
            } else {
                this.Name = name;
                this.DOB = dob;
            }
        }
    }

    export class BirthingUnit 
    {
        // MaxItemsToRetrieve
        private _people: People[];

        constructor() {
            this._people = [];
        }

        // GetPeoples
        // @param j
        // @returns Array<object>
        public GetPeople(i: number): People[] 
        {
            for (let j = 0; j < i; j++) {
                try 
                {
                    // Creates a dandon Name
                    let name: string = "";
                    let random = Math.random();
                    if (Math.floor(Math.random() * 1) == 0) {
                        name = "Bob";
                    } 
                    else { name = "Betty"; }
                    // Adds new people to the list
                    this._people.push(new People(name, new Date(Date.now() - (Math.floor(Math.random() * (85 - 18) + 18) * 356 * 24 * 60 * 60 * 1000))));
                }
                catch (e) {
                    // Dont think this should ever happen
                    throw new Error("Something failed in user creation");
                }
            }
            return this._people;
        }

        private GetBobs(olderThan30: boolean): People[] {
            return olderThan30 ? this._people.filter(x => x.Name == "Bob" && x.DOB >= new Date(Date.now() - (30 * 356 * 24 * 60 * 60 * 1000))) : this._people.filter(x => x.Name == "Bob");
        }

        public GetMarried(p: People, lastName: string): string 
        {
            if (lastName.includes("test"))
                return p.Name;
            if ((p.Name.length + lastName).length > 255) {
                (p.Name + " " + lastName).substring(0, 255);
            }

            return p.Name + " " + lastName;
        }
    }
}